flutter-studio/src/io/flutter/android/AndroidModuleLibraryManager.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2018 The Chromium Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+package io.flutter.android;
+
+import com.android.tools.idea.gradle.project.sync.GradleSyncInvoker;
+import com.android.tools.idea.gradle.project.sync.GradleSyncListener;
+import com.intellij.ProjectTopics;
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
+import com.intellij.openapi.components.ServiceManager;
+import com.intellij.openapi.diagnostic.Logger;
+import com.intellij.openapi.project.DumbService;
+import com.intellij.openapi.project.Project;
+import com.intellij.openapi.project.ProjectManager;
+import com.intellij.openapi.project.impl.ProjectImpl;
+import com.intellij.openapi.roots.ModuleRootEvent;
+import com.intellij.openapi.roots.ModuleRootListener;
+import com.intellij.openapi.roots.OrderRootType;
+import com.intellij.openapi.roots.libraries.Library;
+import com.intellij.openapi.roots.libraries.LibraryTable;
+import com.intellij.openapi.roots.libraries.LibraryTablesRegistrar;
+import com.intellij.openapi.roots.libraries.PersistentLibraryKind;
+import com.intellij.openapi.util.io.FileUtilRt;
+import com.intellij.openapi.vfs.*;
+import com.intellij.util.containers.hash.HashSet;
+import io.flutter.sdk.AbstractLibraryManager;
+import io.flutter.sdk.FlutterSdkUtil;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Arrays;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+
+import static io.flutter.android.AndroidModuleLibraryType.LIBRARY_KIND;
+import static io.flutter.android.AndroidModuleLibraryType.LIBRARY_NAME;
+
+/**
+ * Manages the Android Libraries library, which hooks the libraries used by Android modules referenced in a project
+ * into the project, so full editing support is available.
+ *
+ * @see AndroidModuleLibraryType
+ * @see AndroidModuleLibraryProperties
+ */
+public class AndroidModuleLibraryManager extends AbstractLibraryManager<AndroidModuleLibraryProperties> {
+  private static final Logger LOG = Logger.getInstance(AndroidModuleLibraryManager.class);
+  private static final String BUILD_FILE_NAME = "build.gradle";
+  private final AtomicBoolean isUpdating = new AtomicBoolean(false);
+
+  public AndroidModuleLibraryManager(@NotNull Project project) {
+    super(project);
+  }
+
+  public void update() {
+    doGradleSync(getProject(), (Project x) -> updateAndroidLibraryContent(x));
```

Comment:
I assume you've thought about this, but do we want to drive the gradle sync, or notice when one has happened and then query gradle somewhow?

Reply:
That's a good question, so here's probably more of an answer than you expected. :) Currently, the only way to initiate a Gradle sync is to open the Android project in a new window, which we're trying to get away from. You could do it from the command line, I suppose, but I'm not really interested in that. Eventually we want to have some ability (such as Project Structure) to add dependents to Android modules without launching a new Android Studio for that purpose, and then we'd want to make sure a sync happened once but not twice, but that's in the future. Today, you can edit the Gradle build files (via AS open on a Flutter project) but that won't trigger a sync without this code.